storage/google/cloud/storage/bucket.py

```diff
@@ -829,3 +834,77 @@ def make_public(self, recursive=False, future=False, client=None):
             for blob in blobs:
                 blob.acl.all().grant_read()
                 blob.acl.save(client=client)
+
+    def generate_upload_policy(
+            self, conditions, expiration=None, client=None):
+        """Create a signed upload policy for uploading objects.
+
+        This method generates and signs a policy document. You can use
+        `policy documents`_ to allow visitors to a website to upload files to
+        Google Cloud Storage without giving them direct write access.
+
+        For example:
+
+        .. literalinclude:: storage_snippets.py
+          :start-after: [START policy_document]
+          :end-before: [END policy_document]
+
+        .. _policy documents:
+            https://cloud.google.com/storage/docs/xml-api\
+            /post-object#policydocument
+
+        :type expiration: datetime
+        :param expiration: Optional expiration in UTC. If not specified, the
+            policy will expire in 1 hour.
+
+        :type conditions: list
+        :param conditions: A list of conditions as described in the
+            `policy documents`_ documentation.
+
+        :type client: :class:`~google.cloud.storage.client.Client` or
+                `NoneType``
+        :param client: Optional. The client to use.  If not passed, falls back
+            to the ``client`` stored on the current bucket.
+
+        :rtype: dict
+        :returns: A dictionary of (form field name, form field value) of form
+            fields that should be added to your HTML upload form in order to
+            attach the signature.
+        """
+        client = self._require_client(client)
+        credentials = client._base_connection.credentials
+
+        if not isinstance(credentials, google.auth.credentials.Signing):
+            auth_uri = ('http://google-cloud-python.readthedocs.io/en/latest/'
+                        'google-cloud-auth.html#setting-up-a-service-account')
```

Comment:
`Blob.sign` uses a specialized helper `google.cloud.credentials.generate_signed_url` - which seems pretty weird. Signed URLs are only used by storage afaict, so maybe we should file a bug to replace the core helper with a generic `sign_bytes` and move the url signing logic into storage.